### PR TITLE
Move to rich comparison of backends.

### LINKED ISSUE #211: Use rich comparison to compare backend objects
This is needed when (for example) skipping all in a factory of backends. For example, Dask might implement a backend factory and want to skip ALL of them with a single `skip_backend`.

### PR DESCRIPTION
**Pull request by hameerabbasi**
Fixes #211.

### CODE REVIEW

File: uarray/_uarray_dispatch.cxx [hunk 1]
```diff
@@ -500,7 +500,7 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
       {
         auto it = std::find_if(
           skip.begin(), skip.end(),
-          [&](const py_ref & be) { return be.get() == backend; });
+          [&](const py_ref & be) { return PyObject_RichCompareBool(be.get(), backend, Py_EQ); });
```
Comment (hameerabbasi) [inline uarray/_uarray_dispatch.cxx]: Still needs some error detection here.

File: uarray/_uarray_dispatch.cxx [hunk 2]
```diff
@@ -498,18 +498,36 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
...
-          [&](const py_ref & be) { return be.get() == backend; });
+          [&](const py_ref & be)
+          {
+            int result = PyObject_RichCompareBool(be.get(), backend, Py_EQ);
+            if (result < 0)
+            {
+              success = false;
+              return 0;
+            }
+            return result;
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: How about
```cpp
auto result = PyObject_RichCompareBool(be.get(), backend, Py_EQ);
success = (result >= 0);
return (result != 0);
```
Not only is it a bit more compact, it also breaks out of the `find_if` on error.

File: uarray/_uarray_dispatch.cxx [hunk 3]
```diff
@@ -498,18 +498,36 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
...
-        return (it != skip.end());
+        return (int)(it != skip.end());
       };
 
   LoopReturn ret = LoopReturn::Continue;
   for (int i = pref.size()-1; i >= 0; --i)
   {
     auto options = pref[i];
-    if (should_skip(options.backend))
+    int skip = should_skip(options.backend);
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]: You'll have to update the other uses of `should_skip` as well.

File: uarray/_uarray_dispatch.cxx [hunk 4]
```diff
@@ -498,18 +498,36 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
   auto should_skip =
     [&](PyObject * backend)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
     [&](PyObject * backend) -> int
```

File: uarray/_uarray_dispatch.cxx [hunk 5]
```diff
@@ -498,18 +498,36 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
...
+            return result;
+          });
+
+        if (!success)
+        {
+          return -1;
+        }
 
-        return (it != skip.end());
+        return (int)(it != skip.end());
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
        return (it != skip.end());
```

File: uarray/_uarray_dispatch.cxx [hunk 6]
```diff
@@ -522,7 +536,10 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
 
   auto& globals = global_domain_map[domain_key];
   auto& global_options = globals.global;
-  if (global_options.backend && !should_skip(global_options.backend))
+  int skip_current = global_options.backend.get() != nullptr ? should_skip(global_options.backend) : 1;
+  if (skip_current < 0)
+    return LoopReturn::Error;
+  if (global_options.backend && !skip_current)
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
  if (!skip_current)
```

Comment (peterbell10) [reply uarray/_uarray_dispatch.cxx]: You've already  done the `nullptr` check above.

File: uarray/_uarray_dispatch.cxx [hunk 7]
```diff
@@ -522,7 +536,10 @@ LoopReturn for_each_backend(const std::string & domain_key, Callback call)
 
   auto& globals = global_domain_map[domain_key];
   auto& global_options = globals.global;
-  if (global_options.backend && !should_skip(global_options.backend))
+  int skip_current = global_options.backend.get() != nullptr ? should_skip(global_options.backend) : 1;
```
Comment (peterbell10) [inline uarray/_uarray_dispatch.cxx]:
```suggestion
  int skip_current = global_options.backend ? should_skip(global_options.backend) : 1;
```

**Comment by hameerabbasi** [comment]
### Benchmark

```python
with ua.set_backend(Backend):
    with ua.skip_backend(Backend2):
        %timeit mm()
```

`master`: 394 ns ± 6.04 ns
`rich-compare`: 412 ns ± 5.61 ns

**Comment by peterbell10** [comment]
Is that benchmarking types that don't specialise the comparison operators?

**Comment by hameerabbasi** [comment]
Yes, it is. The base case. Although I measured a backend with a simple `__eq__` method which uses the `is` operator for comparison, it doesn't seem to produce a measurable difference.

**Pull request merged**